### ISSUE TITLE
pgwire: invalid error message

### ISSUE THREAD

**Issue by jethrogb** [cockroachdb/cockroach#19162]
I have now twice encountered a case where Cockroach returns an error result that does not contain the “message primary" (`M`) field. This field is required: https://www.postgresql.org/docs/current/static/libpq-exec.html#libpq-pg-diag-message-primary . My PG driver aborts the program because of this, so I have not been able to ascertain the exact cause.

```
I171010 10:53:50.502371 1 util/log/clog.go:1011  [config] file created at: 2017/10/10 10:53:50
I171010 10:53:50.502371 1 util/log/clog.go:1011  [config] running on machine: node4
I171010 10:53:50.502371 1 util/log/clog.go:1011  [config] binary: CockroachDB CCL v1.0.3 (linux amd64, built 2017/07/06 17:46:06, go1.8.3)
I171010 10:53:50.502371 1 util/log/clog.go:1011  [config] arguments: [/opt/cockroachdb/cockroach start --insecure --store /data/cockroachdb --advertise-host 10.197.66.1 --join 10.197.66.2 --join 10.197.66.3]
```

Here's an excerpt from the logs of this node around the time this happened (approximately 10:58:42):

```
W171010 10:58:02.034868 258 storage/node_liveness.go:291  [n1,hb] slow heartbeat took 1.1s
I171010 10:58:03.458003 196 server/status/runtime.go:225  [n1] runtime stats: 654 MiB RSS, 412 goroutines, 113 MiB/366 MiB/533 MiB GO alloc/idle/total, 78 MiB/105 MiB CGO alloc/total, 2631.73cgo/sec, 0.84/0.05 %(u/s)time, 0.00 %gc (14x)
W171010 10:58:12.689625 114 storage/engine/rocksdb.go:1220  batch [1/53/40] commit took 562.231542ms (>500ms):
goroutine 114 [running]:
runtime/debug.Stack(0x196eb60, 0x2788360, 0xc420014448)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x79
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatch).commitInternal(0xc4264be240, 0xc43e74fc01, 0x1f1, 0x1f1)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1220 +0x14d
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatch).Commit(0xc4264be240, 0x7bea01, 0x2bf5ac0, 0x7fbacbcaacd0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1167 +0x432
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked(0xc42071ef00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2831 +0x533
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequest(0xc420474800, 0x7fbacbcaacd0, 0xc42cb92510, 0xc43bd32900, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3109 +0xcc5
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue(0xc420474800, 0xe1)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3388 +0x172
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc420694ea0, 0xc420699400)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:228 +0x24b
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x7fbacbcaacd0, 0xc420c7a330)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:168 +0x33
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc420c62360, 0xc420699400, 0xc420338820)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:215 +0xf7
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:216 +0xad
W171010 10:58:12.689914 121 storage/store.go:3422  [n1,s1,r259/1:/System/tsd/cr.store.queue.sp…] handle raft ready: 0.7s [processed=0]
W171010 10:58:12.792757 147 storage/store.go:3422  [n1,s1,r106/1:/System/tsd/cr.node.sql.mem.i…] handle raft ready: 0.7s [processed=0]
W171010 10:58:12.792854 143 storage/store.go:3422  [n1,s1,r429/1:/System/tsd/cr.store.{gc…-in…}] handle raft ready: 0.7s [processed=0]
I171010 10:58:13.457532 196 server/status/runtime.go:225  [n1] runtime stats: 657 MiB RSS, 433 goroutines, 146 MiB/334 MiB/533 MiB GO alloc/idle/total, 73 MiB/109 MiB CGO alloc/total, 2505.69cgo/sec, 0.80/0.07 %(u/s)time, 0.00 %gc (12x)
W171010 10:58:13.499494 258 storage/node_liveness.go:291  [n1,hb] slow heartbeat took 3.5s
W171010 10:58:15.281723 152 storage/engine/rocksdb.go:1220  batch [1/53/42] commit took 595.926527ms (>500ms):
goroutine 152 [running]:
runtime/debug.Stack(0x196eb60, 0x2788360, 0xc420014448)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x79
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatch).commitInternal(0xc433f80000, 0xc42381e401, 0x36, 0x400)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1220 +0x14d
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatch).Commit(0xc433f80000, 0x2363f501, 0x2bf5ac0, 0x7fbacbcaacd0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1167 +0x432
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked(0xc420b01500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2831 +0x533
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequest(0xc420474800, 0x7fbacbcaacd0, 0xc426f25440, 0xc438996600, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3109 +0xcc5
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue(0xc420474800, 0xf5)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3388 +0x172
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc420694ea0, 0xc420699400)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:228 +0x24b
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x7fbacbcaacd0, 0xc420c84480)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:168 +0x33
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc420c625c0, 0xc420699400, 0xc420338da0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:215 +0xf7
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:216 +0xad
W171010 10:58:15.768426 258 storage/node_liveness.go:291  [n1,hb] slow heartbeat took 1.3s
W171010 10:58:21.691067 258 storage/node_liveness.go:291  [n1,hb] slow heartbeat took 2.7s
I171010 10:58:23.456869 196 server/status/runtime.go:225  [n1] runtime stats: 659 MiB RSS, 427 goroutines, 134 MiB/347 MiB/533 MiB GO alloc/idle/total, 75 MiB/110 MiB CGO alloc/total, 3122.27cgo/sec, 0.58/0.06 %(u/s)time, 0.00 %gc (13x)
W171010 10:58:23.856004 64304 storage/replica.go:1686  [n1,s1,r453/1:/System/tsd/cr.node.sql.mem.c…] context canceled before command queue: Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, Noop, ... 374 skipped ..., Noop, Noop, Noop, Noop, Noop
W171010 10:58:24.774412 258 storage/node_liveness.go:291  [n1,hb] slow heartbeat took 1.3s
I171010 10:58:33.458302 196 server/status/runtime.go:225  [n1] runtime stats: 662 MiB RSS, 422 goroutines, 95 MiB/384 MiB/533 MiB GO alloc/idle/total, 80 MiB/115 MiB CGO alloc/total, 2741.78cgo/sec, 0.66/0.05 %(u/s)time, 0.00 %gc (12x)
I171010 10:58:43.456839 196 server/status/runtime.go:225  [n1] runtime stats: 665 MiB RSS, 427 goroutines, 109 MiB/371 MiB/533 MiB GO alloc/idle/total, 83 MiB/116 MiB CGO alloc/total, 2629.92cgo/sec, 0.68/0.03 %(u/s)time, 0.00 %gc (10x)
W171010 10:58:52.362531 258 storage/node_liveness.go:291  [n1,hb] slow heartbeat took 1.9s
I171010 10:58:53.457081 195 gossip/gossip.go:448  [n1] gossip status (ok, 3 nodes)
gossip client (0/3 cur/max conns)
gossip server (2/3 cur/max conns, infos 1634/400 sent/received, bytes 283293B/100130B sent/received)
  2: 10.197.66.2:26257 (3m33s)
  3: 10.197.66.3:26257 (1m46s)
I171010 10:58:53.458389 196 server/status/runtime.go:225  [n1] runtime stats: 666 MiB RSS, 430 goroutines, 112 MiB/368 MiB/533 MiB GO alloc/idle/total, 85 MiB/117 MiB CGO alloc/total, 2385.51cgo/sec, 0.53/0.04 %(u/s)time, 0.00 %gc (10x)
```

**Comment by justinj** [comment]
Hi @jethrogb, thanks for the report! I suspect this might have been fixed by #19103, does this reproduce reliably enough that you would be able to try a build off of master and see if you still run into problems? It's possible you were hitting this if there was an incorrect number of type hints in the message you were sending to Cockroach, but I can't be completely sure.

**Comment by jethrogb** [comment]
This was a combination of cockroach segfaulting and the PG driver trying to interpret 0 bytes read as an error response. I filed #19169 for the segfault and will look at the PG driver later (but that shouldn't be your problem).
